common/main/cpp/native_glue.cc

```diff
@@ -168,14 +171,37 @@ namespace litecore::jni {
     // ----------------------------------------------------------------------------
     // JVM
     // ----------------------------------------------------------------------------
-    JavaVM *gJVM;
 
-    int attachCurrentThread(JNIEnv **env) {
+    int attachJVM(JNIEnv **env, const char *caller) {
+        jint ret = gJVM->GetEnv(reinterpret_cast<void **>(env), JNI_VERSION_1_6);
+
+        if (ret == JNI_OK)
+            return JNI_OK;
+
+        if (ret != JNI_EDETACHED) {
+            jniLog("%s: Failed to get the JVM environment: %d", caller, ret);
+            return ret;
+        }
+
+        ret =
 #ifdef JNI_VERSION_1_8
-        return gJVM->AttachCurrentThread(reinterpret_cast<void **>(env), NULL);
+                gJVM->AttachCurrentThread(reinterpret_cast<void **>(env), NULL);
 #else
-        return gJVM->AttachCurrentThread(env, nullptr);
+                gJVM->AttachCurrentThread(env, nullptr);
 #endif
+
+        if (ret == JNI_OK)
+            return JNI_EDETACHED;
```

Comment:
This is confusing to me. Doesn't `JNI_EDETACHED` mean not attached? Why not returning `JNI_OK`?

Reply:
If I read this right, and this probably deserves a comment because of its unusualness, JNI_EDETACHED is being used here to indicate that the state was detached before the call to attach (possibly to do or not do cleanup later)

Reply:
I feel that the return here is inconsistent with the return in line 179. Basically if it's already attached when calling GetEnv, it returns `JNI_OK` in line 179 but it returns `JNI_EDETACHED` in line 194 if the attach result is successful.

Reply:
I kind of get it now. Agree with Jim that it deserves comment as it's an unusual return. I understand that the returned status is the result of the operation but what it returns now is either the previous state or the action that should be done next.

Reply:
Jim is correct.  I will add a comment.